js/indicators/tema.src.js

```diff
@@ -0,0 +1,235 @@
+                      1,
+                      EMApercent,
+                      calTEMA,
+                      -1,
+                      SMA
+                    )[1];
+                    TEMAPoint = [
+                        xVal[i - 3],
+                        correctFloat(3 * ema - 3 * emaEMA + emaEMAEMA)
+                    ];
+                    TEMA.push(TEMAPoint);
+                    xDataTema.push(TEMAPoint[0]);
+                    yDataTema.push(TEMAPoint[1]);
+                    calTEMA = emaEMAEMA;
+                }
+            }
+
+            // Calculate the last ema and emaEMA points
+            ema = EmaArr[i - period - 1];
+            emaEMA = EMA.prototype.calculateEma(
+              [],
+              [ema],
+              1,
+              EMApercent,
+              calDEMA,
+              -1,
+              SMA
+            )[1];
+            emaEMAArr.push(emaEMA);
+
+            // Calculate the last tema points
+            ema = EmaArr[i - period - 2];
+            emaEMA = emaEMAArr[i - 2 * period - 1];
+            emaEMAEMA = EMA.prototype.calculateEma(
+              [],
+              [emaEMA],
+              1,
+              EMApercent,
+              calTEMA,
+              -1,
+              SMA
+            )[1];
+            TEMAPoint = [
+                xVal[i - 3],
+                correctFloat(3 * ema - 3 * emaEMA + emaEMAEMA)
+            ];
+            TEMA.push(TEMAPoint);
+            xDataTema.push(TEMAPoint[0]);
+            yDataTema.push(TEMAPoint[1]);
+            calTEMA = emaEMAEMA;
+
+            i++;
+
+            ema = EmaArr[i - period - 2];
+            emaEMA = emaEMAArr[i - 2 * period - 1];
+            emaEMAEMA = EMA.prototype.calculateEma(
+              [],
+              [emaEMA],
+              1,
+              EMApercent,
+              calTEMA,
+              -1,
+              SMA
+            )[1];
+            TEMAPoint = [
+                xVal[i - 3],
+                correctFloat(3 * ema - 3 * emaEMA + emaEMAEMA)
+            ];
+            TEMA.push(TEMAPoint);
+            xDataTema.push(TEMAPoint[0]);
+            yDataTema.push(TEMAPoint[1]);
+
+            return {
+                values: TEMA,
+                xData: xDataTema,
+                yData: yDataTema
+            };
+        }
+    });
+
+/**
+ * A `EMA` series. If the [type](#series.ema.type) option is not
+ * specified, it is inherited from [chart.type](#chart.type).
+ *
+ * @type {Object}
+ * @since 7.0.0
+ * @extends series,plotOptions.ema
+ * @excluding
+ *          allAreas,colorAxis,compare,compareBase,data,dataParser,dataURL,
+ *          joinBy,keys,stacking,showInNavigator,navigatorOptions,pointInterval,
+ *          pointIntervalUnit,pointPlacement,pointRange,pointStart,joinBy
+ * @product highstock
+ * @apioption series.tema
+ */
+
+/**
+ * @type {Array<Object|Array>}
+ * @since 7.0.0
+ * @extends series.sma.data
+ * @product highstock
+ * @apioption series.tema.data
+ */
```

Comment:
In many places these lines are used:

```
 emaEMAEMA = EMA.prototype.calculateEma(
                      [],
                      [emaEMA],
                      1,
                      EMApercent,
                      calTEMA,
                      -1,
                      SMA
                    )[1];
```

```
            ema = EmaArr[i - period - 2];
            emaEMA = emaEMAArr[i - 2 * period - 1];
            emaEMAEMA = EMA.prototype.calculateEma(
              [],
              [emaEMA],
              1,
              EMApercent,
              calTEMA,
              -1,
              SMA
            )[1];
            TEMAPoint = [
                xVal[i - 3],
                correctFloat(3 * ema - 3 * emaEMA + emaEMAEMA)
            ];
            TEMA.push(TEMAPoint);
            xDataTema.push(TEMAPoint[0]);
            yDataTema.push(TEMAPoint[1]);
```

Only one parameter is different. Consider to extract it as parametrized function or optimize it.